opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,676 @@
+  declare_parameter("filter_coef", 0.1);
+  declare_parameter("desired_distance", 1.0);
+  declare_parameter("skip_orientation", true);
+  declare_parameter("search_by_rotating", false);
+  declare_parameter("odom_topic", "odom");
+  declare_parameter("transform_tolerance", 0.1);
+}
+
+nav2::CallbackReturn
+FollowingServer::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring %s", get_name());
+  auto node = shared_from_this();
+
+  get_parameter("controller_frequency", controller_frequency_);
+  get_parameter("detection_timeout", detection_timeout_);
+  get_parameter("rotate_to_object_timeout", rotate_to_object_timeout_);
+  get_parameter("static_object_timeout", static_object_timeout_);
+  get_parameter("linear_tolerance", linear_tolerance_);
+  get_parameter("angular_tolerance", angular_tolerance_);
+  get_parameter("max_retries", max_retries_);
+  get_parameter("base_frame", base_frame_);
+  get_parameter("fixed_frame", fixed_frame_);
+  get_parameter("allow_backward", allow_backward_);
+  get_parameter("desired_distance", desired_distance_);
+  get_parameter("skip_orientation", skip_orientation_);
+  get_parameter("search_by_rotating", search_by_rotating_);
+  get_parameter("transform_tolerance", transform_tolerance_);
+  RCLCPP_INFO(get_logger(), "Controller frequency set to %.4fHz", controller_frequency_);
+
+  vel_publisher_ = std::make_unique<nav2_util::TwistPublisher>(node, "cmd_vel");
+  tf2_buffer_ = std::make_shared<tf2_ros::Buffer>(node->get_clock());
+
+  // Create odom subscriber for backward blind docking
+  std::string odom_topic;
+  get_parameter("odom_topic", odom_topic);
+  odom_sub_ = std::make_unique<nav_2d_utils::OdomSubscriber>(node, odom_topic);
+
+  // Create the action server for dynamic following
+  following_action_server_ = node->create_action_server<FollowObject>(
+    "follow_object",
+    std::bind(&FollowingServer::followObject, this),
+    nullptr, std::chrono::milliseconds(500),
+    true);
+
+  // Create composed utilities
+  controller_ =
+    std::make_unique<opennav_docking::Controller>(node, tf2_buffer_, fixed_frame_, base_frame_);
+
+  // Setup filter
+  double filter_coef;
+  get_parameter("filter_coef", filter_coef);
+  filter_ = std::make_unique<opennav_docking::PoseFilter>(filter_coef, detection_timeout_);
+
+  // And publish the filtered pose for debugging
+  filtered_dynamic_pose_pub_ =
+    create_publisher<geometry_msgs::msg::PoseStamped>("filtered_dynamic_pose");
+
+  // Initialize static object detection variables
+  static_timer_initialized_ = false;
+  static_object_start_time_ = rclcpp::Time(0);
+
+  return nav2::CallbackReturn::SUCCESS;
+}
+
+nav2::CallbackReturn
+FollowingServer::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating %s", get_name());
+
+  auto node = shared_from_this();
+
+  tf2_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf2_buffer_);
```

Comment:
Note: there's a PR open right now to have this use the API for `this, true` to spin an executor using the server's node. https://github.com/ros-navigation/navigation2/pull/5406 Not merged yet, but something to keep in mind if it does get merged soon